pkg/resources/external_volume.go

```diff
@@ -0,0 +1,717 @@
+		Description:      blocklistedCharactersFieldDescription("Identifier for the external volume; must be unique for your account."),
+		DiffSuppressFunc: suppressIdentifierQuoting,
+	},
+	// A list is used as the order of storage locations matter. Storage location position in the list is used to select
+	// the active storage location - https://docs.snowflake.com/en/user-guide/tables-iceberg-storage#active-storage-location
+	// This is also why it has been left as one list with optional cloud dependent parameters, rather than splitting into
+	// one list per cloud provider.
+	"storage_location": {
+		Type:        schema.TypeList,
+		Required:    true,
+		MinItems:    1,
+		Description: "List of named cloud storage locations in different regions and, optionally, cloud platforms. Minimum 1 required. Note that not all parameter combinations are valid as they depend on the given storage_provider. Consult [the docs](https://docs.snowflake.com/en/sql-reference/sql/create-external-volume#cloud-provider-parameters-cloudproviderparams) for more details on this.",
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"storage_location_name": {
+					Type:        schema.TypeString,
+					Required:    true,
+					Description: "Name of the storage location. Must be unique for the external volume.",
+				},
+				"storage_provider": {
+					Type:             schema.TypeString,
+					Required:         true,
+					ValidateDiagFunc: sdkValidation(sdk.ToStorageProvider),
+					DiffSuppressFunc: SuppressIfAny(NormalizeAndCompare(sdk.ToStorageProvider)),
+					Description:      fmt.Sprintf("Specifies the cloud storage provider that stores your data files. Valid values are (case-insensitive): %s.", possibleValuesListed(sdk.ValidStorageProviderString)),
+				},
+				"storage_base_url": {
+					Type:        schema.TypeString,
+					Required:    true,
+					Description: "Specifies the base URL for your cloud storage location.",
+				},
+				"storage_aws_role_arn": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					Description: "Specifies the case-sensitive Amazon Resource Name (ARN) of the AWS identity and access management (IAM) role that grants privileges on the S3 bucket containing your data files.",
+				},
+				"storage_aws_external_id": {
+					Type:        schema.TypeString,
+					Computed:    true,
+					Description: "External ID that Snowflake uses to establish a trust relationship with AWS.",
+				},
+				"encryption_type": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					Description: "Specifies the encryption type used.",
+					DiffSuppressFunc: func(k, oldValue, newValue string, d *schema.ResourceData) bool {
+						return oldValue == "NONE" && newValue == ""
+					},
+				},
+				"encryption_kms_key_id": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					Description: "Specifies the ID for the KMS-managed key used to encrypt files.",
+				},
+				"azure_tenant_id": {
```

Comment:
Would you have an example on how I could set that up in the resource?

There are a few requirements to consider:
1. The storage location list needs to be ordered for all storage providers, so we can only have 1 list, not 3. Reason is explained here - https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/90f2c88bee1c2388860fd763b217dfe1dbf7fa4b/pkg/resources/external_volume.go#L26.
2. Each element in the list should only specify one of [“s3”, “gcs”, “azure”]. However, `ConflictsWith` doesn’t work with lists I believe - https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3106#discussion_r1799021583.

Reply:
Sorry for not being clear here. We can do this on our side, since it requires a bit more design and implementation. We can have something like 
```
resource "snowflake_external_volume" "complete" {
  name         = var.name
  comment      = var.comment
  allow_writes = var.allow_writes

  storage_location {
    s3 {
        storage_location_name = ...
        storage_aws_external_id = ...
    }
  }

  storage_location {
    azure {
        storage_location_name = ...
        azure_tenant_id = ...
    }
  }
}
```

1. We would still still have one list of storage locations. We'd use pseudo-lists with `MaxItems: 1` to specify the `s3`, `gcp` and `azure` blocks.
2. `MaxItems` would be 1, so this would work. We do this in other resources.

As I wrote, this requires more work and the design you proposed is really good. We can have a team discussion about this later.

Reply:
Sounds good, happy for you to take care of it 🙂 Two things we should do after that has been done:

- Remove the custom 'conflictswith' checks in `extractStorageLocations` e.g. https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/8ac4af1d5241357819bd2ed7a4a7f6f993fd1bf2/pkg/resources/external_volume.go#L477
- Add this test - https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/8ac4af1d5241357819bd2ed7a4a7f6f993fd1bf2/pkg/resources/external_volume_acceptance_test.go#L2555